### ISSUE TITLE
NVDA doesn't read text typed into most on-line code interpreters

### ISSUE THREAD

**Issue by jiwi12** [nvaccess/nvda#7285]
I am an assistive technologist working with blind children in public schools. My agency would like to find a way for our students to participate in coding clubs with their sighted peers. Many clubs use on-line coding tutorials that have built-in interpreters in which students type and run sample programs. These are favored over downloadable code editors because most school computers are heavily restricted, so the purely on-line learning environment is often the only one available. I have been testing various websites with NVDA for accessibility, but I've found only one that works so far.

I am running Windows 7 Professional, Service Pack 1, with an installed copy of NVDA 2017.2, and the behavior below occurs in both Firefox 54.0 and Internet Explorer 11.

When I go to the [homepage](https://quorumlanguage.com/) of the Quorum programming language website and enter code into the interpreter, I can arrow through it and edit it as in most text fields. (It's under the level 2 heading labeled, "Try Quorum!"

However, on all the other code-learning websites I've tried, I can't tell if the interpreters are accepting text because NVDA says "blank" when I arrow left and right. See [W3Schools](https://www.w3schools.com/html/tryit.asp?filename=tryhtml_elements) for an example.

I don't know enough about HTML to discern the differences between these two text fields; without looking at the source code, they both seem to me like garden-variety edit boxes, but one is accessible and one isn't. I'd appreciate  help figuring out whether my second example or any similar sites, (such as Codecademy, Kahn Academy, etc.), can be coded so students can practice various programming languages on-line, or if this lack of output can be addressed within NVDA. (Incidentally, Jaws 18 exhibits the same behavior.)

**Comment by jcsteh** [comment]
Unfortunately, most code editors don't use standard HTML text editing facilities. They have a textarea to accept input from the user, but it's not actually used to display the text. Once you type or paste something, it gets removed from the textarea. The text is displayed via other means (non-editable HTML/CSS or something else). While you can often read previously entered text by switching to browse mode, you can't read it when focused in the textarea. It's not possible to make this accessible; there's no standard way for screen readers to know how these two things are connected.

The only way to make editable content accessible is to use HTML textarea or contentEditable. [CodeMirror](https://codemirror.net/), which is a project used by a lot of code editors, now has a contentEditable mode. (See the inputStyle option in the [Configuration section of the CodeMirror manual](https://codemirror.net/doc/manual.html#config) for details.) While not the default yet, authors can choose it and it's being considered as the default for the next major version of CodeMirror. This should get us some way towards improving things, though quite a few browser bugs need to be fixed for this to work well.

Closing as can't fix, since there's nothing we can do on our side.
